[DATA-3338] fix-stability-of-vision-capture-all-from-camera-and-camera-get-images

nicksanford opened the pull request:
Why:
1. the collectors for vision.CaptureAllFromCamera and camera.GetImages write the responses from those methods as tabular data. That results a go object being allocated for every pixle of the images in those method's responses. At high capture rates that results in heap fragmentation and viam-server crashing due to running out of memory. See https://viam.atlassian.net/browse/DATA-3191 for the tests that show that.
2. To fix that problem we need all binary collectors to write binary results. This requires allowing a binary capture file to contain multiple binary results (as that is what camera.GetImages returns). 
3. We also needed to write annotations to capture files as that is what vision.CaptureAllFromCamera needs to describe what the ML model detected in the image.

Fixes:
- [Fix stability of vision.CaptureAllFromCamera](https://viam.atlassian.net/browse/DATA-3338)
- [Fix stability of camera.GetImages](https://viam.atlassian.net/browse/DATA-3339)
- [GetImages capture files fail to sync if a single image is under the 256 kb capture file size limit](https://viam.atlassian.net/browse/DATA-3418)

Changes:
1. Change signature of collectors from `(ctx context.Context, _ map[string]*anypb.Any) (interface{}, error)` to `(ctx context.Context, _ map[string]*anypb.Any) (data.CaptureResult, error)` see [data/collector_types.go](https://github.com/viamrobotics/rdk/pull/4557/files#diff-67c0823aa24d3667ac603603cf934d7c0d988547e19a88cd2d48b3aa1337dc99R22) for more info
5. Change CaptureFuncs to determine the capture request and response time rather than data/collector.go
6. Change all tabular collectors to use `data.NewTabularCaptureResult` to create tabular capture results.
7. Update collector tests to assert on multiple sensor data values
8. Update data sync to accept binary capture files with multiple readings (as in the case of camera.GetImages)
9. Update data sync to handle both legacy camera.GetImages capture files and the new, more efficient ones
10. Update data sync to set both the datasyncpb.MimeType and the datasyncpb.UploadMetadata.FileExtension fields
11. Update data sync to move capture file to failed directory in cases where the capture file contains unrecoverable data (such as multiple capture types within a given capture file, or the capture file header type being something other than binary or tabular)

nicksanford commented on components/camera/collectors_test.go:
```diff
@@ -94,59 +88,50 @@ func TestCollectors(t *testing.T) {
 			name:      "ReadImage returns a non nil binary response",
 			collector: camera.NewReadImageCollector,
-			expected: &datasyncpb.SensorData{
+			expected: []*datasyncpb.SensorData{{
```
add mime_type to tests.

nicksanford commented on data/capture_file.go:
```diff
@@ -30,9 +29,10 @@ const (
+	GetImages            = "GetImages"
+	nextPointCloud       = "NextPointCloud"
+	pointCloudMap        = "PointCloudMap"
+	captureAllFromCamera = "CaptureAllFromCamera"
```
added `captureAllFromCamera` to binary methods

nicksanford commented on data/capture_file.go:
```diff
@@ -240,49 +240,6 @@ func getFileTimestampName() string {
-}
-
-// GetFileExt gets the file extension for a capture file.
-func GetFileExt(dataType v1.DataType, methodName string, parameters map[string]string) string {
```
moved here: https://github.com/viamrobotics/rdk/pull/4557/files#diff-67c0823aa24d3667ac603603cf934d7c0d988547e19a88cd2d48b3aa1337dc99R366

nicksanford commented on data/capture_file.go:
```diff
@@ -240,49 +240,6 @@ func getFileTimestampName() string {
 }
 
-// TODO DATA-246: Implement this in some more robust, programmatic way.
-func getDataType(methodName string) v1.DataType {
```
moved here: https://github.com/viamrobotics/rdk/pull/4557/files#diff-67c0823aa24d3667ac603603cf934d7c0d988547e19a88cd2d48b3aa1337dc99R193

With the viaonCaptureAllFromCamera added to the binary data list.

nicksanford commented on services/datamanager/builtin/sync/exponential_retry.go:
```diff
@@ -168,6 +168,14 @@ func getNextWait(lastWait time.Duration, isOffline bool) time.Duration {
+		return true
+	}
+
+	for _, e := range terminalCaptureFileErrs {
```
https://github.com/viamrobotics/rdk/pull/4557/files#diff-d4e29d2325615faf57196f0792e0f204e4a16c442bcc44924cfc35675321ba7aR25-R29
This ensures that we give up if the capture file is invalid.

dmhilly commented on data/collector.go:
```diff
@@ -178,10 +183,27 @@ func (c *collector) tickerBasedCapture(started chan struct{}) {
 	}
 }
 
+func (c *collector) validateReadingType(t CaptureType) error {
```
At a high level, why do both the collector and the reading it returns have a CaptureType? It seems we never expect them to be different, so why do we encode this information in two different places?

nicksanford replied:
Great question!

The definition of that is a tabular or binary collector lives [here](https://github.com/viamrobotics/rdk/pull/4557/files#diff-67c0823aa24d3667ac603603cf934d7c0d988547e19a88cd2d48b3aa1337dc99R193) and is called [here](https://github.com/viamrobotics/rdk/pull/4557/files#diff-943a1876bce4656fb8ce485d913b26670b3449d0d430a98002fe1c2439373975R220) which is called [here](https://github.com/viamrobotics/rdk/pull/4557/files#diff-b38086d606e0e0fd934d9795d68af0065d61bfd9b5ea4a960aa69ba9f511615cR327) in capture.go.

That means that (as is also true on main) datamanager/capture/capture.go is responsible for determining whether a given resource/method pair is binary or tabular. 
 
However [data.CaptureFunc](https://github.com/viamrobotics/rdk/pull/4557/files/f3beccf39a296f5d51c4759a83b3dfa1dd59b336#diff-67c0823aa24d3667ac603603cf934d7c0d988547e19a88cd2d48b3aa1337dc99R22) (the interface method that a collector implements) returns a CaptureResult, which can either be binary or tabular. validateReadingType checks that the response (what the collector actually returned) matches the expectations of collector.go.

This is a risk that exists today in main, I'm just now checking for the mismatch.

The only way I can see to resolve this runtime check is to create 2 different constructors collectors can implement, one which returns a binary response and another that returns a tabular response. That would be cleaner, but would take more time for me to implement.

Let me know if that makes sense.

dmhilly replied:
Great explanation, thank you so much! I'm happy with the choice you've made to just add validation for now. I like your idea for how to catch this at compile time, we can always do that in the future. This PR is pretty big already :)

nicksanford replied:
https://github.com/viamrobotics/rdk/blob/main/services/datamanager/builtin/capture/capture.go#L327
which calls https://github.com/viamrobotics/rdk/blob/main/data/capture_file.go#L220
which calls https://github.com/viamrobotics/rdk/blob/main/data/capture_file.go#L244


The response of data.BuildCaptureMetadata is passed to the collector constructor here: https://github.com/viamrobotics/rdk/blob/main/services/datamanager/builtin/capture/capture.go#L345
which determines the datasync.DataType of the entire capture file: https://github.com/viamrobotics/api/blob/main/proto/viam/app/datasync/v1/data_sync.proto#L134

Nothing in main enforces that the DataCaptureMetadata.Type must match the type of the resource/method pair response.

dmhilly replied:
Thanks for linking all of that! LGTM!

dmhilly reviewed (changes requested):
Really nice work Nick! I haven't finished reviewing this and will take another look on Monday, but figured I'd publish the feedback that I have ready now.

A request: it seems like there are a lot of changes going on here that do things other than fix the stability of capture all from camera. The PR description was super useful for helping me understand _what_ changed here, but not _why_. Mind adding those details in the PR description and updating the PR's title?

nicksanford commented on components/arm/collectors.go:
```diff
@@ -39,18 +40,20 @@ func newEndPositionCollector(resource interface{}, params data.CollectorParams)
 		}
 		o := v.Orientation().OrientationVectorDegrees()
-		return pb.GetEndPositionResponse{
+		return data.NewTabularCaptureResult(timeRequested, pb.GetEndPositionResponse{
```
camera.GetImages was created for cameras like the intel realsense, which returns multiple images (such as color and depth) from a single point in time. That time is determined by a clock on the camera and returned in the camera.GetImages response in the captured_at field. 

Data capture stores camera.GetImages responses are currently tabular data, which is a big stability problem. See https://viam.atlassian.net/browse/DATA-3418.

Data sync currently syncs camera.GetImages responses as binary data (there is a special case in sync.go go for camera.GetImage capture files) with the requested_at and received_at timestamp set to the camera.GetImages.captured_at timestamp from the camera's clock.

To remove this special casing in the sync path, I decided that making the requested_at and received_at timestamps a collector responsibility was the best way to maintain the current behavior without requiring generic capture and sync to have knowledge of the camera.GetImages special case.

dmhilly replied:
Thanks for the explanation, SGTM!

nicksanford commented on data/collector_types.go:
```diff
@@ -0,0 +1,394 @@
+	return CaptureResult{
+		Timestamps: Timestamps{
+			TimeRequested: reqT,
+			TimeReceived:  time.Now(),
```
The reason for the difference is that we never need to override time_requested  and time_received for tabular data. 

We only need to override binary data timestamps in the case of camera.GetImages: https://github.com/viamrobotics/rdk/pull/4557#discussion_r1866316226

n0nick replied:
I agree with @dmhilly in terms of interface symmetry / readability

dmhilly replied:
Makes sense. I still found the difference in the function signatures to be confusing while reading the code, but that's just me. I know this is a big PR already, so up to you.

nicksanford replied:
done

nicksanford commented on data/collector_types.go:
```diff
@@ -0,0 +1,394 @@
+	defaultFileExt := ""
+	switch dataType {
+	case CaptureTypeTabular:
+		return ".dat"
```
These are not used elsewhere. This code was moved from https://github.com/viamrobotics/rdk/pull/4557/files/f3beccf39a296f5d51c4759a83b3dfa1dd59b336#r1854115109

That said, I'll make them constants.

nicksanford commented on components/camera/collectors_test.go:
```diff
@@ -94,59 +88,52 @@ func TestCollectors(t *testing.T) {
 			camera: cam,
 		},
 		{
 			name:      "GetImages returns a non nil tabular response",
```
good point, will update

nicksanford replied:
done

n0nick commented on data/collector_types.go:
```diff
@@ -0,0 +1,394 @@
+		return MimeTypeImageJpeg
+	case camerapb.Format_FORMAT_PNG:
+		return MimeTypeImagePng
+	case camerapb.Format_FORMAT_RAW_RGBA:
+		fallthrough
+	case camerapb.Format_FORMAT_RAW_DEPTH:
+		fallthrough
```
`TODO` comments for tracking these?  or if desired, explain why

nicksanford replied:
how is that?

dmhilly commented on data/collector_types.go:
```diff
@@ -0,0 +1,394 @@
+}
+
+// ToProto converts Annotations to *dataPB.Annotations.
+func (mt Annotations) ToProto() *dataPB.Annotations {
```
Thanks for explaining, that rationale makes sense to me. Will keep in mind in the future.

n0nick replied:
+1 re decoupling API and internal representations. 

re code duplication, in Java Google solves this with [AutoValue](https://github.com/google/auto/blob/main/value/userguide/index.md), I wonder if there's an equivalent pattern in Go (e.g. using `go generate`)

n0nick commented on data/collector_types.go:
```diff
@@ -0,0 +1,394 @@
+// Validate returns an error if the *CaptureResult is invalid.
+func (cr *CaptureResult) Validate() error {
+	var ts Timestamps
+	if cr.Timestamps.TimeRequested == ts.TimeRequested {
```
I found these comparisons (to the unset `ts.Time*`) unintuitive; perhaps compare to a `var emptyTime time.Time` instead for readability?

nicksanford replied:
how is that?

n0nick commented on data/collector_types.go:
```diff
@@ -0,0 +1,394 @@
+type MimeType int
+
+// This follows the mime types supported in
+// https://github.com/viamrobotics/api/pull/571/files#diff-b77927298d8d5d5228beeea47bd0860d9b322b4f3ef45e129bc238ec17704826R75
```
nit: link to https://github.com/viamrobotics/api/blob/d7436a969cbc03bf7e84bb456b6dbfeb51f664d7/proto/viam/app/datasync/v1/data_sync.proto#L69 instead?

nicksanford replied:
done

n0nick reviewed (commented):
I am not done (36 / 41 files reviewed) but going home for the day and figured I'd submit my comments so far. Nothing major. Change is well structured and tested, thanks for that. To be continued!

dmhilly commented on data/collector_types.go:
```diff
@@ -0,0 +1,394 @@
+		TabularData: TabularData{
+			Payload: &structpb.Struct{
+				Fields: map[string]*structpb.Value{
+					"readings": structpb.NewStructValue(&structpb.Struct{Fields: values}),
```
Ah thanks - okay I vaguely remember this now.

nicksanford replied:
I've added a comment that both reflects what the author believed when this code was originally written and what we now know with the benefit of hindsight and more understanding of how proto works.

dmhilly replied:
Hah, love the comment. Thanks Nick! 🙌

nicksanford commented on data/collector_types.go:
```diff
@@ -0,0 +1,394 @@
+}
+
+// GetDataType returns the DataType of the method.
+func GetDataType(methodName string) CaptureType {
```
changed to `MethodToCaptureType`

nicksanford commented on components/movementsensor/collectors.go:
added back

nicksanford commented on services/datamanager/builtin/builtin_sync_test.go:
```diff
@@ -343,7 +344,7 @@ func TestDataCaptureUploadIntegration(t *testing.T) {
 						extra map[string]interface{},
 					) (spatialmath.Pose, error) {
-						return spatialmath.NewZeroPose(), nil
+						return spatialmath.NewPoseFromPoint(r3.Vector{X: 1, Y: 2, Z: 3}), nil
```
because the zero pose is going to result in an all zero proto value. I wanted to confirm that it works with non zero values. It does.

nicksanford commented on data/collector_test.go:
```diff
@@ -297,13 +320,14 @@ func TestLogErrorsOnlyOnce(t *testing.T) {
 	mockClock.Add(interval * 5)
 
-	close(wrote)
+	// close(wrote)
```
fixed

nicksanford commented on components/camera/collectors.go:
```diff
@@ -112,14 +123,26 @@ func newReadImageCollector(resource interface{}, params data.CollectorParams) (d
+		}
+
+		actualOutMIME, _ := utils.CheckLazyMIMEType(mimeStr.Value)
+		// NOTE (Nick S): This is the mime type that the user described in the config, not the actual
```
no longer relevant given other changes that went into main ahead of this PR.

nicksanford commented on services/datamanager/builtin/sync/upload_data_capture_file.go:
```diff
@@ -2,22 +2,32 @@ package sync
+	errMultipleReadingTypes                   = errors.New("sensor readings contain multiple types")
+	errSensorDataTypesDontMatchUploadMetadata = errors.New("sensor reaings types don't match upload metadata")
+	errInvalidCaptureFileType                 = errors.New("invalid capture file type")
+	terminalCaptureFileErrs                   = []error{
```
done

nicksanford commented on data/collector.go:
```diff
@@ -292,8 +281,24 @@ func (c *collector) writeCaptureResults() {
 		case msg := <-c.captureResults:
-			if err := c.target.Write(msg); err != nil {
-				c.logger.Error(errors.Wrap(err, fmt.Sprintf("failed to write to collector %s", c.target.Path())).Error())
+			proto := msg.ToProto()
```
It could have. It was less code change to have them both take a slice of sensor readings given the old `Write` method took a single sensor reading.

dmhilly replied:
SGTM, thanks for explaining.

nicksanford commented on data/collector_types.go:
```diff
@@ -0,0 +1,394 @@
+// ToProto converts a CaptureResult into a []*datasyncPB.SensorData{}.
+func (cr *CaptureResult) ToProto() []*datasyncPB.SensorData {
+	ts := cr.Timestamps
+	if td := cr.TabularData.Payload; td != nil {
```
There should not be a case where both `TabularData` and `Binaries` are both empty. Vaalidate() would have returned an error if that were the case.

The expectation is that CaptureResult.ToProto() is only called if CaptureResult.Validate() passed

nicksanford commented on data/capture_buffer.go:
```diff
@@ -76,7 +103,26 @@ func (b *CaptureBuffer) Write(item *v1.SensorData) error {
 
-	return b.nextFile.WriteNext(item)
+	for _, item := range items {
+		if err := b.nextFile.WriteNext(item); err != nil {
```
for tabular data it will still only be one sensor data at a time so no this isn't a regression

dmhilly replied:
Thanks!

nicksanford commented on services/datamanager/builtin/sync/upload_data_capture_file.go:
```diff
@@ -41,63 +51,105 @@ func uploadDataCaptureFile(ctx context.Context, f *data.CaptureFile, conn cloudC
-	return &v1.UploadMetadata{
+func checkUploadMetadaTypeMatchesSensorDataType(md *datasyncPB.DataCaptureMetadata, sensorDataTypeSet map[data.CaptureType]struct{}) error {
+	var captureType data.CaptureType
+	// get first element of single element set
+	for k := range sensorDataTypeSet {
```
Someone in sol eng or external to viam writes a program that writes capture files and they put both binary and tabular data in that file. That or we have a bad bug.

This is not an expected condition but we are running this check so that if it does happen, we move the file to the failed directory and move on.

nicksanford commented on services/datamanager/builtin/sync/upload_data_capture_file.go:
```diff
@@ -125,55 +180,161 @@ func getImagesTimestamps(res *pb.GetImagesResponse, sensorData *v1.SensorData) (
+	// we only go down this path if the capture method returned multiple binary
+	// responses, which at time of writing, only includes camera.GetImages data.
+	for i, sd := range sensorData {
+		// Otherwise use the unary endpoint
```
done

nicksanford commented on services/datamanager/builtin/sync/upload_data_capture_file.go:
```diff
@@ -125,55 +180,161 @@ func getImagesTimestamps(res *pb.GetImagesResponse, sensorData *v1.SensorData) (
+		// between calls if the data in the request struct changes
+		clonedMD := proto.Clone(uploadMD).(*datasyncPB.UploadMetadata)
+		if err := uploadBinarySensorData(ctx, client, clonedMD, sd); err != nil {
+			return err
```
that is possible. If that happens we will retry both images if the error we get is transient. 

Worst case we will duplicate the first image.

That is also the case today.

nicksanford commented on services/datamanager/builtin/sync/upload_data_capture_file.go:
```diff
@@ -125,55 +180,161 @@ func getImagesTimestamps(res *pb.GetImagesResponse, sensorData *v1.SensorData) (
-		// Then call the function to send the rest.
-		if err := sendStreamingDCRequests(ctx, c, toUpload.GetBinary(), path, logger); err != nil {
-			return errors.Wrap(err, "StreamingDataCaptureUpload failed to sync")
+	for i, sd := range sensorData {
```
This is a good point. This problem is that we want to check if our data sync request is above or below the limit, which, to your point, isn't the same thing as the capture file size. 

It is also not the same thing as the size of the sensorData turned into proto (as that doesn't contain the upload metadata).

Yes I would prefer to use streaming endpoints for all uploads, which would require an API change as we don't have that endpoint today for tabular data.

It is also a good idea to change to a different wire format for tabular data (maybe bison) as otherwise, you can't predict the size of the proto data without serializing it twice.

I do think this is best left for a later change.

nicksanford commented on services/datamanager/builtin/sync/upload_data_capture_file.go:
```diff
@@ -193,8 +354,8 @@ func sendStreamingDCRequests(
 			chunk := contents[i:end]
```
exactly!

nicksanford commented on services/datamanager/builtin/sync/upload_data_capture_file.go:
```diff
@@ -211,17 +372,32 @@ func sendStreamingDCRequests(
+	case cameraPB.Format_FORMAT_PNG:
 		return ".png"
-	case pb.Format_FORMAT_RAW_DEPTH:
+	case cameraPB.Format_FORMAT_RAW_DEPTH:
```
I don't think so... `cameraPB` is an alias to `go.viam.com/api/component/camera/v1` which is what `pb` was also an alias to in the old version... I just changed it b/c I needed to be able to refer to `datasyncPB` as well

dmhilly replied:
That sounds good. Saw the ticket come in and the change for 1 elsewhere in this PR. My only concern is whether it's a good idea to add information as an image tag/annotation that the user did not explicitly ask us to include. Seems like a new pattern - passing information to the user through a tag - that might be unexpected for users. I feel like additional parameters would be more appropriate although we don't have a way to dynamically add these params at capture time. Maybe this is fine, but wanted to call it out.

nicksanford replied:
I don't think there is much difference between the source name of each camera and a classification from an ML model. In either case it is a string that contains metadata about the image. IMO for any resource method response that contains both tabular and binary data we will need to have some place to put the tabular data when we sync that response to the datasync api. I think the user IS asking to be able to distinguish which camera sensor each image in the response came from when they query their data in binary data by filter.

Also, for the intel realsense (the main reason why this method exists) the `souce_name` strings are configured by the user: https://github.com/viam-modules/viam-camera-realsense in the robot config:
 the `sensors` list becomes the source_names in the GetImages response.

dmhilly replied:
Yeah that's fair. I guess CaptureAllFromCamera feels different to me because a classification is typically an annotation on an image, whereas us returning this metadata as a tag might be unexpected. But given this is the only avenue we have at the moment for returning arbitrary metadata along with the image, I'm okay with this for now.

nicksanford commented:
> Really nice work Nick! I haven't finished reviewing this and will take another look on Monday, but figured I'd publish the feedback that I have ready now.
> 
> A request: it seems like there are a lot of changes going on here that do things other than fix the stability of capture all from camera. The PR description was super useful for helping me understand _what_ changed here, but not _why_. Mind adding those details in the PR description and updating the PR's title?

Everything in this PR is in service to the "why" section I've added to the PR description. I've updated the PR description to include that this is also fixing camera.GetImages.

Is that enough detail on why?

dmhilly reviewed (changes requested):
Just a couple more questions - I'm almost done reviewing, just have to look at one more test file.

dmhilly commented:
> Everything in this PR is in service to the "why" section I've added to the PR description. I've updated the PR description to include that this is also fixing camera.GetImages.
>
> Is that enough detail on why?

Yes! Thank you Nick!

dmhilly commented on components/camera/collectors.go:
```diff
@@ -157,8 +157,9 @@ func newGetImagesCollector(resource interface{}, params data.CollectorParams) (d
 			binaries = append(binaries, data.Binary{
-				Payload:  imgBytes,
-				MimeType: data.CameraFormatToMimeType(format),
+				Annotations: data.Annotations{Classifications: []data.Classification{{Label: img.SourceName}}},
```
Hey, we implemented dynamic tagging! 😆

nicksanford replied:
For binary data yes!

dmhilly reviewed (commented):
Left just one follow up comment re: communicating to users the source of each image returned by GetImages.

dmhilly reviewed (approved):
LGTM! Great work on this Nick 🎉🎉

n0nick reviewed (approved):
👍

nicksanford commented on services/datamanager/builtin/sync/upload_data_capture_file_test.go:
```diff
@@ -0,0 +1,893 @@
+	partID := "my-part-id"
+
+	now := time.Now()
+	ts := data.Timestamps{TimeRequested: now, TimeReceived: time.Now()}
```
no b/c in general `TimeRequested` happens before `TimeReceived`. camera.GetImages is the only exception
